File: skillet/src/main/java/dev/kkorolyov/pancake/skillet/Skillet.java
package dev.kkorolyov.pancake.skillet;

import dev.kkorolyov.pancake.skillet.model.GenericComponent;
import dev.kkorolyov.pancake.skillet.model.Workspace;
import dev.kkorolyov.pancake.skillet.ui.component.ComponentList;
import dev.kkorolyov.pancake.skillet.ui.entity.EntityList;
import dev.kkorolyov.pancake.skillet.ui.entity.EntityTabPane;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.Alert;
import javafx.scene.control.Alert.AlertType;
import javafx.scene.control.Menu;
import javafx.scene.control.MenuBar;
import javafx.scene.control.MenuItem;
import javafx.scene.image.Image;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.BorderPane;
import javafx.stage.FileChooser;
import javafx.stage.FileChooser.ExtensionFilter;
import javafx.stage.Stage;
import java.io.File;
import java.util.List;
import java.util.stream.Collectors;

import static dev.kkorolyov.pancake.skillet.decorator.UIDecorator.decorate;

public class Skillet extends Application {
	private static final String WORKSPACE_FILE_EXTENSION = ".mfn";

	private final Workspace workspace = new Workspace();
	private final ResourceHandler resourceHandler = new ResourceHandler();

	private Stage stage;

	private final EntityList entityList = new EntityList(workspace);
	private final ComponentList componentList = new ComponentList(workspace.getComponentFactory());
	private final EntityTabPane entityTabPane = new EntityTabPane();

	public static void main(String[] args) {
		Application.launch(args);
	}

	@Override
	public void start(Stage primaryStage) {
		stage = primaryStage;

		workspace.register(entityTabPane);

		entityTabPane
				.onEntitySelected(selected -> {
					workspace.setActiveEntity(selected);
					componentList.setEntity(selected);
				});

		componentList.onComponentSelected(component ->
				workspace.getActiveEntity()
						.ifPresent(entity -> entity.addComponent(component))
		);

		addDefaultComponents();

		primaryStage.setTitle("Skillet - Pancake Design Workspace");
		primaryStage.getIcons().add(new Image("pancake-icon.png"));

		BorderPane root = new BorderPane();

		// Menu bar
		root.setTop(new MenuBar(
				new Menu("_Entity", null,
						decorate(new MenuItem("_New"))
								.action(this::newEntity)
								.get(),
						decorate(new MenuItem("_Save"))
								.action(this::saveWorkspace)
								.get(),
						decorate(new MenuItem("_Open"))
								.action(this::loadWorkspace)
								.get()),
				new Menu("_Components", null,
						decorate(new MenuItem("_Load"))
								.action(this::loadComponents)
								.get())));

		Scene scene = new Scene(root, 480, 480);
		scene.getStylesheets().add("style.css");

		// Shortcuts
		scene.addEventHandler(KeyEvent.KEY_PRESSED, e -> {
			if (e.isControlDown()) {
				switch (e.getCode()) {
					case N:
						newEntity();
						break;
					case S:
						saveWorkspace();
						break;
					case O:
						loadWorkspace();
						break;
					case L:
						loadComponents();
						break;
				}
			}
		});

		root.setLeft(entityList.getRoot());
		root.setRight(componentList.getRoot());
		root.setCenter(entityTabPane.getRoot());

		primaryStage.setScene(scene);
		primaryStage.show();
	}
	private void addDefaultComponents() {
		workspace.getComponentFactory().add(
				resourceHandler.load("defaults" + WORKSPACE_FILE_EXTENSION).getComponents(),
				false);
	}

	private void newEntity() {
		workspace.newEntity();
	}
	private void saveWorkspace() {
		FileChooser chooser = buildFileChooser();
		chooser.setTitle("Save Workspace");
		chooser.setInitialFileName("workspace" + WORKSPACE_FILE_EXTENSION);

		File result = chooser.showSaveDialog(stage);
		if (result != null) resourceHandler.save(workspace, result.toString());
	}
	private void loadWorkspace() {
		FileChooser chooser = buildFileChooser();
		chooser.setTitle("Load Workspace");

		File result = chooser.showOpenDialog(stage);
		if (result != null) workspace.addWorkspace(resourceHandler.load(result.toString()));
	}

	private void loadComponents() {
		FileChooser chooser = buildFileChooser();
		chooser.setTitle("Load Components");

		List<File> results = chooser.showOpenMultipleDialog(stage);
		if (results != null) {
			List<GenericComponent> addedComponents = results.stream()
					.map(file -> resourceHandler.load(file.toString()))	// To workspaces
					.flatMap(Workspace::streamComponents)
					.collect(Collectors.toList());

			workspace.getComponentFactory().add(addedComponents, false);

			Alert addedComponentsInfo = new Alert(
					AlertType.INFORMATION,
					addedComponents.stream()
							.map(component -> component.getName() + " - " + component.getAttributes().size() + " attributes")
							.collect(Collectors.joining(System.lineSeparator())));
			addedComponentsInfo.setTitle("Added Components");
			addedComponentsInfo.setHeaderText(addedComponentsInfo.getTitle());

			addedComponentsInfo.show();
		}
	}

	private FileChooser buildFileChooser() {
		FileChooser chooser = new FileChooser();
		chooser.getExtensionFilters().add(new ExtensionFilter("Skillet Workspaces", "*" + WORKSPACE_FILE_EXTENSION));
		chooser.setInitialDirectory(new File("").getAbsoluteFile());

		return chooser;
	}
}

File: skillet/src/main/kotlin/dev/kkorolyov/pancake/skillet/Skillet.kt
package dev.kkorolyov.pancake.skillet

import dev.kkorolyov.pancake.skillet.model.Workspace
import dev.kkorolyov.pancake.skillet.ui.component.ComponentList
import dev.kkorolyov.pancake.skillet.ui.entity.EntityList
import dev.kkorolyov.pancake.skillet.ui.entity.EntityTabPane
import javafx.application.Application
import javafx.scene.Scene
import javafx.scene.control.Alert
import javafx.scene.control.Menu
import javafx.scene.control.MenuBar
import javafx.scene.image.Image
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.BorderPane
import javafx.stage.FileChooser.ExtensionFilter
import javafx.stage.Stage

fun main(vararg args: String) {
	Application.launch(Skillet::class.java, *args)
}

/**
 * Skillet application entry point.
 */
class Skillet : Application() {
	companion object {
		private const val WORKSPACE_FILE_EXTENSION = ".mfn"
	}
	private val workspace: Workspace = Workspace()

	private val entityList: EntityList = EntityList(workspace)
	private val componentList: ComponentList = ComponentList(workspace.componentFactory)
	private val entityTabPane: EntityTabPane = EntityTabPane(workspace,
			entitySelected = { componentList.refreshComponents(it) })

	private lateinit var stage: Stage

	override fun start(primaryStage: Stage?) {
		stage = primaryStage ?: throw NullPointerException()
		ResourceHandler.stage = stage

		workspace.componentFactory.add(
				ResourceHandler.load("defaults$WORKSPACE_FILE_EXTENSION").components,
				false)

		stage.title = "Skillet - Pancake Design Workspace"
		stage.icons.add(Image("pancake-icon.png"))

		val root = BorderPane().apply {
			top = MenuBar(
					Menu("_Entity").apply {
						item("_New", { newEntity() })
						item("_Save", { saveWorkspace() })
						item("_Open", { loadWorkspace() })
					},
					Menu("_Components").apply {
						item("_Load") { loadComponents() }
					})
			left = entityList.root
			right = componentList.root
			center = entityTabPane.root
		}
		stage.scene = Scene(root, 480.0, 480.0)
		stage.scene.stylesheets.add("style.css")
		stage.scene.addEventHandler(KeyEvent.KEY_PRESSED) { e ->
			if (e.isControlDown) {
				when (e.code) {
					KeyCode.N -> newEntity()
					KeyCode.S -> saveWorkspace()
					KeyCode.O -> loadWorkspace()
					KeyCode.L -> loadComponents()
					else -> {}
				}
			}
		}
		stage.show()
	}

	private fun newEntity() {
		workspace.newEntity()
	}
	private fun saveWorkspace() {
		ResourceHandler.chooseSave("Save Workspace", "workspace$WORKSPACE_FILE_EXTENSION", ExtensionFilter("Skillet Workspaces", "*$WORKSPACE_FILE_EXTENSION"))?.let {
			ResourceHandler.save(workspace, it.toString())
		}
	}
	private fun loadWorkspace() {
		ResourceHandler.chooseLoad("Load Workspace", extensionFilters = ExtensionFilter("Skillet Workspaces", "*$WORKSPACE_FILE_EXTENSION"))?.let {
			workspace.addWorkspace(ResourceHandler.load(it.toString()))
		}
	}
	private fun loadComponents() {
		ResourceHandler.chooseLoadMultiple("Load Components", ExtensionFilter("Components", "*$WORKSPACE_FILE_EXTENSION"))?.let {
			val addedComponents = it
					.map { file -> ResourceHandler.load(file.toString()) }
					.flatMap(Workspace::components)

			workspace.componentFactory.add(addedComponents, false)

			val addedComponentsInfo = Alert(
					Alert.AlertType.INFORMATION,
					addedComponents.joinToString(System.lineSeparator()) { "${it.name} - ${it.attributes.size} attributes" })
			addedComponentsInfo.title = "Added components"
			addedComponentsInfo.headerText = addedComponentsInfo.title

			addedComponentsInfo.show()
		}
	}
}
